File: src/bun.js/test/expect/toHaveLength.zig
pub fn toHaveLength(
    this: *Expect,
    globalThis: *JSGlobalObject,
    callframe: *CallFrame,
) bun.JSError!JSValue {
    defer this.postMatch(globalThis);
    const thisValue = callframe.this();
    const arguments_ = callframe.arguments_old(1);
    const arguments = arguments_.slice();

    if (arguments.len < 1) {
        return globalThis.throwInvalidArguments("toHaveLength() takes 1 argument", .{});
    }

    this.incrementExpectCallCounter();

    const expected: JSValue = arguments[0];
    const value: JSValue = try this.getValue(globalThis, thisValue, "toHaveLength", "<green>expected<r>");

    if (!value.isObject() and !value.isString()) {
        var fmt = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        return globalThis.throw("Received value does not have a length property: {f}", .{value.toFmt(&fmt)});
    }

    if (!expected.isNumber()) {
        var fmt = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        return globalThis.throw("Expected value must be a non-negative integer: {f}", .{expected.toFmt(&fmt)});
    }

    const expected_length: f64 = expected.asNumber();
    if (@round(expected_length) != expected_length or std.math.isInf(expected_length) or std.math.isNan(expected_length) or expected_length < 0) {
        var fmt = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        return globalThis.throw("Expected value must be a non-negative integer: {f}", .{expected.toFmt(&fmt)});
    }

    const not = this.flags.not;
    var pass = false;

    const actual_length = try value.getLengthIfPropertyExistsInternal(globalThis);

    if (actual_length == std.math.inf(f64)) {
        var fmt = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
        return globalThis.throw("Received value does not have a length property: {f}", .{value.toFmt(&fmt)});
    } else if (std.math.isNan(actual_length)) {
        return globalThis.throw("Received value has non-number length property: {}", .{actual_length});
    }

    if (actual_length == expected_length) {
        pass = true;
    }

    if (not) pass = !pass;
    if (pass) return .js_undefined;

    // handle failure
    if (not) {
        const expected_line = "Expected length: not <green>{d}<r>\n";
        const signature = comptime getSignature("toHaveLength", "<green>expected<r>", true);
        return this.throw(globalThis, signature, "\n\n" ++ expected_line, .{expected_length});
    }

    const expected_line = "Expected length: <green>{d}<r>\n";
    const received_line = "Received length: <red>{d}<r>\n";
    const signature = comptime getSignature("toHaveLength", "<green>expected<r>", false);
    return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ expected_length, actual_length });
}

const bun = @import("bun");
const std = @import("std");

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toHaveLength.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
#[allow(unused_imports)] use super::{JSValueTestExt, JSGlobalObjectTestExt, BigIntCompare, make_formatter};
use bun_jsc::console_object::Formatter;

use super::Expect;
use super::get_signature;

// TODO(port): #[bun_jsc::host_fn(method)] — must be inside `impl Expect`; shim wired by JsClass codegen
pub fn to_have_length(
    this: &Expect,
    global: &JSGlobalObject,
    frame: &CallFrame,
) -> JsResult<JSValue> {
    let (this, value, not) =
        this.matcher_prelude(global, frame.this(), "toHaveLength", "<green>expected<r>")?;

    let arguments_ = frame.arguments_old::<1>();
    let arguments = arguments_.slice();

    if arguments.len() < 1 {
        return Err(global.throw_invalid_arguments(format_args!("toHaveLength() takes 1 argument")));
    }

    let expected: JSValue = arguments[0];

    if !value.is_object() && !value.is_string() {
        let mut fmt = super::make_formatter(global);
        return Err(global.throw(format_args!(
            "Received value does not have a length property: {}",
            value.to_fmt(&mut fmt),
        )));
    }

    if !expected.is_number() {
        let mut fmt = super::make_formatter(global);
        return Err(global.throw(format_args!(
            "Expected value must be a non-negative integer: {}",
            expected.to_fmt(&mut fmt),
        )));
    }

    let expected_length: f64 = expected.as_number();
    if expected_length.round() != expected_length
        || expected_length.is_infinite()
        || expected_length.is_nan()
        || expected_length < 0.0
    {
        let mut fmt = super::make_formatter(global);
        return Err(global.throw(format_args!(
            "Expected value must be a non-negative integer: {}",
            expected.to_fmt(&mut fmt),
        )));
    }

    let mut pass = false;

    let actual_length = value.get_length_if_property_exists_internal(global)?;

    if actual_length == f64::INFINITY {
        let mut fmt = super::make_formatter(global);
        return Err(global.throw(format_args!(
            "Received value does not have a length property: {}",
            value.to_fmt(&mut fmt),
        )));
    } else if actual_length.is_nan() {
        return Err(global.throw(format_args!(
            "Received value has non-number length property: {}",
            actual_length,
        )));
    }

    if actual_length == expected_length {
        pass = true;
    }

    if not {
        pass = !pass;
    }
    if pass {
        return Ok(JSValue::UNDEFINED);
    }

    // handle failure
    if not {
        // PERF(port): was comptime getSignature — const fn evaluated at compile time
        let signature: &str = get_signature("toHaveLength", "<green>expected<r>", true);
        return this.throw(
            global,
            signature,
            format_args!("\n\nExpected length: not <green>{}<r>\n", expected_length),
        );
    }

    // PERF(port): was comptime getSignature — const fn evaluated at compile time
    let signature: &str = get_signature("toHaveLength", "<green>expected<r>", false);
    this.throw(
        global,
        signature,
        format_args!(
            "\n\nExpected length: <green>{}<r>\nReceived length: <red>{}<r>\n",
            expected_length, actual_length,
        ),
    )
}

// ported from: src/test_runner/expect/toHaveLength.zig
